geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/SlotAdvisor.java

```diff
@@ -0,0 +1,230 @@
+ * the License.
+ */
+
+package org.apache.geode.redis.internal;
+
+import static org.apache.geode.redis.internal.RegionProvider.REDIS_REGION_BUCKETS;
+import static org.apache.geode.redis.internal.RegionProvider.REDIS_SLOTS_PER_BUCKET;
+import static org.apache.geode.redis.internal.cluster.RedisMemberInfoRetrievalFunction.RedisMemberInfo;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.logging.log4j.Logger;
+
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.execute.FunctionService;
+import org.apache.geode.cache.execute.ResultCollector;
+import org.apache.geode.cache.partition.PartitionMemberInfo;
+import org.apache.geode.cache.partition.PartitionRegionHelper;
+import org.apache.geode.cache.partition.PartitionRegionInfo;
+import org.apache.geode.distributed.DistributedMember;
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.redis.internal.cluster.RedisMemberInfoRetrievalFunction;
+import org.apache.geode.redis.internal.data.RedisData;
+import org.apache.geode.redis.internal.data.RedisKey;
+
+public class SlotAdvisor {
+
+  private static final Logger logger = LogService.getLogger();
+
+  /**
+   * Mapping buckets to slot information
+   */
+  private final List<MemberBucketSlot> memberBucketSlots;
+
+  /**
+   * Cache of member ids to member IP address and redis listening port
+   */
+  private final Map<String, Pair<String, Integer>> hostPorts = new HashMap<>();
+  private final PartitionedRegion dataRegion;
+  private final Region<String, Object> configRegion;
+
+  SlotAdvisor(Region<RedisKey, RedisData> dataRegion, Region<String, Object> configRegion) {
+    this.dataRegion = (PartitionedRegion) dataRegion;
+    this.configRegion = configRegion;
+    memberBucketSlots = new ArrayList<>(RegionProvider.REDIS_REGION_BUCKETS);
+    for (int i = 0; i < REDIS_REGION_BUCKETS; i++) {
+      memberBucketSlots.add(null);
+    }
+  }
+
+  public boolean isLocal(RedisKey key) {
+    return dataRegion.getRegionAdvisor().getBucket(key.getBucketId()).getBucketAdvisor()
+        .isPrimary();
+  }
+
+  public Pair<String, Integer> getHostAndPortForKey(RedisKey key) {
+    int bucketId = key.getBucketId();
+    MemberBucketSlot mbs = updateBucketDetails(bucketId);
+
+    return Pair.of(mbs.getPrimaryIpAddress(), mbs.getPrimaryPort());
+  }
+
+  public Map<String, List<Integer>> getMemberBuckets() {
+    initializeBucketsIfNecessary();
+
+    Map<String, List<Integer>> memberBuckets = new HashMap<>();
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      String memberId =
+          dataRegion.getRegionAdvisor().getBucketAdvisor(bucketId).getPrimary().getUniqueId();
```

Comment:
Maybe it would be better to keep member ids as `InternalDistributedMember` in most places until we actually return a string to the user? I normally don't like converting objects to strings and doing comparisons on the strings.

Reply:
Done.